samza-api/src/main/java/org/apache/samza/context/ApplicationContainerContextFactory.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.context;
+
+import java.io.Serializable;
+
+
+/**
+ * An application should implement this if it has a {@link ApplicationContainerContext} that is needed for
+ * initialization. This will be used to create instance(s) of that {@link ApplicationContainerContext}.
+ * <p>
+ * This is {@link Serializable} because it is specified in {@link org.apache.samza.application.ApplicationDescriptor}.
+ * @param <T> concrete type of {@link ApplicationContainerContext} returned by this factory
+ */
+public interface ApplicationContainerContextFactory<T extends ApplicationContainerContext> extends Serializable {
+  /**
+   * Create an instance of the application-defined {@link ApplicationContainerContext}.
+   *
+   * @param jobContext framework-provided job context used for building {@link ApplicationContainerContext}
+   * @param containerContext framework-provided container context used for building {@link ApplicationContainerContext}
+   * @return new instance of the application-defined {@link ApplicationContainerContext}
+   */
+  T create(JobContext jobContext, ContainerContext containerContext);
```

Comment:
Is the idea to pass in all available / existing context objects that ApplicationContainerContext may depend on in the input parameters list? Would that be harder to extend if we need to add a ProcessorContext later?

Reply:
Talked in-person about this:
It would be harder to extend, but the alternative of having a wrapper object which contains these arguments complicates the public API.
With this proposed API, the strategy for extending would be to add a new method which includes a new object and to deprecate the old method.